docs(tip-1091): add enshrined ZoneFactory TIP (#6787)
tempoxyz/tempo: tips/tip-1091.md

bytecode that points at one central `ZonePortal` logic implementation.

## Reference Artifacts

The reference artifacts for this TIP are:

1. [`tips/verify/src/zones/interfaces/IZone.sol`](verify/src/zones/interfaces/IZone.sol)
2. [`tips/verify/src/zones/tempo/ZoneFactory.sol`](verify/src/zones/tempo/ZoneFactory.sol)
3. [`tips/verify/src/zones/tempo/ZonePortal.sol`](verify/src/zones/tempo/ZonePortal.sol)

## Specification

Tempo adds a `ZoneFactory` precompile at:

```solidity
address constant ZONE_FACTORY_ADDRESS = 0x5AF2000000000000000000000000000000000000;
```

The precompile exposes `IZoneFactory`. Its registry behavior MUST match the
canonical Zones implementation: monotonically increasing zone IDs, `zones`,
`zoneCount`, `isZonePortal`, `isValidVerifier`, `verifier`, `messenger`, and
`ZoneCreated`.

Each zone portal is installed at:

```text
ZONE_PORTAL_PREFIX = 0x5AD000000000000000000000
portal = bytes20(ZONE_PORTAL_PREFIX || uint64(zoneId))
```

The suffix is the zone ID encoded big-endian into the low 8 bytes. `zoneId == 0`
is reserved, so `zoneId == 1` maps to
`0x5AD0000000000000000000000000000000000001`.

The central portal logic bytecode is stored at:

```solidity

> address constant ZONE_PORTAL_LOGIC_ADDRESS = 0x5AD1000000000000000000000000000000000000;

Reviewer: Etching the bytecode for this will be a separate TIP once the logic is ready?

Reviewer 2: why should it be a separate TIP?

Author: Yes should be the same TIP, addressed in https://github.com/tempoxyz/tempo/commit/b70e00f75cfaee6137d439e9e773c1801e888fd5

```

Each portal account is etched with this delegatecall proxy runtime:

> ```text

Reviewer: Can we add a solidity mock for this? Would make it more readable

Author: this is the canonical EIP 1167 bytecode - https://eips.ethereum.org/EIPS/eip-1167
will add it to the TIP too

Author: addressed in https://github.com/tempoxyz/tempo/pull/6787/commits/af565952aa7e0cdab6df662579fcc756355d5499

portal_proxy_runtime =
  hex"363d3d373d3d3d363d73"
  || bytes20(ZONE_PORTAL_LOGIC_ADDRESS)
  || hex"5af43d82803e903d91602b57fd5bf3"
```

`createZone` MUST NOT deploy the portal with EVM `CREATE` or `CREATE2`, and
MUST NOT rely on nonce-based address prediction. Instead, the native factory:

1. validates the request as the canonical factory would;
2. assigns the next zone ID;
3. computes the portal vanity address;

> 4. verifies that account has no existing code, storage, or EIP-7702 delegation;

Reviewer 3: nit: Very unlikely but since this is called out in the spec as a check that is run, what would the expected behavior be if the address had existing code?

Reviewer: This should never happen. Should we just drop this check and always etch?

Author: Addressed in https://github.com/tempoxyz/tempo/commit/78559b93147e1ea0847a2cbffce14358d826666f

5. etches `portal_proxy_runtime` at that address;
6. initializes portal state as if the `ZonePortal.sol` constructor ran with the
   `createZone` parameters;
7. records the zone metadata, marks the portal in `isZonePortal`, and emits

>    `ZoneCreated`.

Reviewer 3: Why do we need `isZonePortal` if all portals have a prefixed address and we know the latest `zoneId`?

Reviewer: I think all checks for things like `isZonePortal` should go through a function and not check by address. Checks by address are error-prone and have poor forward compatibility.
(But the implementation can be the address check + comparison with `zoneId`)

Author: Addressed in https://github.com/tempoxyz/tempo/commit/ea45b1be6b3589dd38ed25bc1919a6b44c6f5be8

```solidity
address constant ZONE_FACTORY_ADDRESS = 0x5AF2000000000000000000000000000000000000;
```

The precompile exposes `IZoneFactory`. Its registry behavior MUST match the
canonical Zones implementation: monotonically increasing zone IDs, `zones`,
`zoneCount`, `isZonePortal`, `isValidVerifier`, `verifier`, `messenger`, and
`ZoneCreated`.

Each zone portal is installed at:

```text
ZONE_PORTAL_PREFIX = 0x5AD000000000000000000000
portal = bytes20(ZONE_PORTAL_PREFIX || uint64(zoneId))
```

The suffix is the zone ID encoded big-endian into the low 8 bytes. `zoneId == 0`
is reserved, so `zoneId == 1` maps to
`0x5AD0000000000000000000000000000000000001`.

The central portal logic bytecode is stored at:

```solidity
address constant ZONE_PORTAL_LOGIC_ADDRESS = 0x5AD1000000000000000000000000000000000000;
```

Each portal account is etched with this delegatecall proxy runtime:

```text
portal_proxy_runtime =
  hex"363d3d373d3d3d363d73"
  || bytes20(ZONE_PORTAL_LOGIC_ADDRESS)
  || hex"5af43d82803e903d91602b57fd5bf3"
```

`createZone` MUST NOT deploy the portal with EVM `CREATE` or `CREATE2`, and
MUST NOT rely on nonce-based address prediction. Instead, the native factory:

> 1. validates the request as the canonical factory would;

Reviewer 2: verify caller is an authorized zone deployer

Author: updated here - https://github.com/tempoxyz/tempo/pull/6787/commits/92fe6b2ed5d3076ea62447f2bf56f49f2c360861

---
id: TIP-1091
title: Enshrined ZoneFactory
description: Enshrine ZoneFactory as a Tempo precompile and define native ZonePortal account installation.
authors: Tanishk Goyal
status: Draft
related: Tempo Zones, TIP-20, TIP-403

> protocolVersion: TBD

Reviewer 4: ```suggestion
protocolVersion: T9
```

The portal implementation, verifier, and shared messenger use the following
protocol-managed addresses:

```solidity
address constant ZONE_PORTAL_IMPL_ADDRESS = 0x5AD1000000000000000000000000000000000000;
address constant ZONE_VERIFIER_ADDRESS = 0x5a56000000000000000000000000000000000000;
address constant ZONE_MESSENGER_ADDRESS = 0x5A4d000000000000000000000000000000000000;
```

Each portal account is etched with the standard
[ERC-1167](https://eips.ethereum.org/EIPS/eip-1167) deployed proxy bytecode,
with `ZONE_PORTAL_IMPL_ADDRESS` as the implementation address:

```text
portal_proxy_runtime =
  hex"363d3d373d3d3d363d73"
  || bytes20(ZONE_PORTAL_IMPL_ADDRESS)
  || hex"5af43d82803e903d91602b57fd5bf3"
```

### Initial Factory Configuration

The native factory MUST start with zone ID `1`. Its initial owner is configured
by T9 and remains TBD. The owner MAY transfer ownership to any address, including
the zero address. The shared messenger at `ZONE_MESSENGER_ADDRESS` MUST authorize
`ZONE_FACTORY_ADDRESS`.

### Shared Runtime Management

`setPortalImplementation(address source)` MUST be callable only by the factory
owner. After authorization, its only source validation MUST be

> `EXTCODEHASH(source) != 0`; the factory MUST NOT otherwise validate `source` or

Reviewer 5: ```suggestion
`EXTCODEHASH(source) != EMPTY_CODE_HASH`; the factory MUST NOT otherwise validate `source` or
```

its runtime bytecode. A successful call MUST copy the complete runtime bytecode
at `source` to
`ZONE_PORTAL_IMPL_ADDRESS`, equivalent to `EXTCODECOPY`, and emit
`PortalImplementationUpdated(source, codeHash)` with the observed
`EXTCODEHASH(source)`.

`setZoneMessengerImplementation(address source)` and
`setVerifierImplementation(address source)` MUST have the same authorization,

> sole `EXTCODEHASH(source) != 0` validation, and bytecode-copy semantics, targeting

Reviewer 5: ```suggestion
sole `EXTCODEHASH(source) != EMPTY_CODE_HASH` validation, and bytecode-copy semantics, targeting
```
